File: app/src/main/java/me/instabattle/app/activities/MapActivity.java
package me.instabattle.app.activities;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.support.annotation.NonNull;
import android.support.v4.app.FragmentActivity;
import android.os.Bundle;
import android.support.v4.content.ContextCompat;
import android.util.Log;
import android.view.View;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.UiSettings;
import com.google.android.gms.maps.model.Circle;
import com.google.android.gms.maps.model.CircleOptions;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Marker;
import com.google.android.gms.maps.model.MarkerOptions;

import java.util.HashMap;
import java.util.List;

import me.instabattle.app.models.Battle;
import me.instabattle.app.managers.BattleManager;
import me.instabattle.app.R;
import me.instabattle.app.services.LocationService;
import me.instabattle.app.settings.State;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MapActivity extends FragmentActivity implements OnMapReadyCallback {

    private static final String TAG = "MapActivity";

    public static final float DEFAULT_ZOOM = 14;
    public static final LatLng DEFAULT_VIEW_POINT = new LatLng(59.930969, 30.352445);

    private GoogleMap googleMap;

    private HashMap<String, Battle> battleByMarker;
    private HashMap<String, Circle> circleByMarker;

    private Circle visibleCircle = null;

    public static Class<?> gotHereFrom;

    public static LatLng viewPoint = DEFAULT_VIEW_POINT;
    public static float viewZoom = DEFAULT_ZOOM;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_map);

        SupportMapFragment mapFragment = (SupportMapFragment) getSupportFragmentManager().findFragmentById(R.id.map);
        mapFragment.getMapAsync(this);

        battleByMarker = new HashMap<>();
        circleByMarker = new HashMap<>();
    }


    @Override
    public void onMapReady(GoogleMap googleMap) {
        Log.d(TAG, "map ready");

        this.googleMap = googleMap;

        if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION)
                == PackageManager.PERMISSION_GRANTED) {
            googleMap.setMyLocationEnabled(true);
        } else {
            LocationService.askLocationPermission(this);
        }

        UiSettings settings = googleMap.getUiSettings();
        settings.setZoomControlsEnabled(true);
        settings.setMapToolbarEnabled(false);
        settings.setMyLocationButtonEnabled(true);

        googleMap.moveCamera(CameraUpdateFactory.newLatLngZoom(viewPoint, viewZoom));

        googleMap.setOnInfoWindowClickListener(new GoogleMap.OnInfoWindowClickListener() {
            @Override
            public void onInfoWindowClick(Marker marker) {
                State.chosenBattle = battleByMarker.get(marker.getId());
                BattleActivity.gotHereFrom = MapActivity.class;
                Intent viewBattle = new Intent(MapActivity.this, BattleActivity.class);
                startActivity(viewBattle);
            }
        });

        googleMap.setOnMapClickListener(new GoogleMap.OnMapClickListener() {
            @Override
            public void onMapClick(LatLng latLng) {
                if (visibleCircle != null) {
                    visibleCircle.setVisible(false);
                    visibleCircle = null;
                }
            }
        });

        googleMap.setOnMarkerClickListener(new GoogleMap.OnMarkerClickListener() {
            @Override
            public boolean onMarkerClick(Marker marker) {
                if (visibleCircle != null) {
                    visibleCircle.setVisible(false);
                }
                visibleCircle = circleByMarker.get(marker.getId());
                visibleCircle.setVisible(true);
                return false;
            }
        });

        placeMarkers();
    }

    private void placeMarkers() {
        BattleManager.getAllAndDo(new Callback<List<Battle>>() {
            @Override
            public void onResponse(Call<List<Battle>> call, Response<List<Battle>> response) {
                List<Battle> battles = response.body();
                for (Battle nearBattle : battles) {
                    Marker newMarker = MapActivity.this.googleMap.addMarker(new MarkerOptions()
                            .position(nearBattle.getLocation())
                            .title(nearBattle.getName())
                            .snippet(nearBattle.getEntriesCount() + " photos"));
                    battleByMarker.put(newMarker.getId(), nearBattle);

                    Circle newCircle = MapActivity.this.googleMap.addCircle(new CircleOptions()
                            .center(nearBattle.getLocation())
                            .radius(nearBattle.getRadius())
                            .strokeColor(Color.GREEN)
                            .strokeWidth(4)
                            .visible(false));
                    circleByMarker.put(newMarker.getId(), newCircle);
                }
                Log.d(TAG, "placed markers for " + battles.size() + " battles");
            }

            @Override
            public void onFailure(Call<List<Battle>> call, Throwable t) {
                //TODO
                Log.e(TAG, "cant get battles: " + t);
            }
        });
    }

    @Override
    public void onPause() {
        super.onPause();
        viewPoint = googleMap.getCameraPosition().target;
        viewZoom = googleMap.getCameraPosition().zoom;
    }

    public void goCreateBattle(View v) {
        State.creatingBattle = true;
        Intent createBattle = new Intent(this, CreateBattleActivity.class);
        startActivity(createBattle);
    }

    @Override
    public void onBackPressed() {
        Intent intent = new Intent(this, gotHereFrom);
        startActivity(intent);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String permissions[], @NonNull int[] grantResults) {
        if (requestCode == LocationService.REQUEST_LOCATION_PERMISSION) {
            if (grantResults.length != 1 || grantResults[0] != PackageManager.PERMISSION_GRANTED) {
                //TODO: cant participate message
                Log.d(TAG, "didn't get location permission");
            } else {
                Log.d(TAG, "got location permission");

                try {
                    googleMap.setMyLocationEnabled(true);
                } catch (SecurityException e) {
                    //try catch block is required by API 23 and higher
                    //never get here
                }
            }
        } else {
            super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        }
    }
}

File: app/src/main/kotlin/me/instabattle/app/activities/MapActivity.kt
package me.instabattle.app.activities

import android.Manifest
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Color
import android.support.v4.app.FragmentActivity
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.view.View

import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.Circle
import com.google.android.gms.maps.model.CircleOptions
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions

import java.util.HashMap

import me.instabattle.app.models.Battle
import me.instabattle.app.managers.BattleManager
import me.instabattle.app.R
import me.instabattle.app.services.LocationService
import me.instabattle.app.settings.State
import me.instabattle.app.utils.Utils
import org.jetbrains.anko.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class MapActivity : FragmentActivity(), OnMapReadyCallback, AnkoLogger {
    private lateinit var googleMap: GoogleMap

    private lateinit var battleByMarker: HashMap<String, Battle>
    private lateinit var circleByMarker: HashMap<String, Circle>

    private var visibleCircle: Circle? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_map)

        val mapFragment = supportFragmentManager.findFragmentById(R.id.map) as SupportMapFragment
        mapFragment.getMapAsync(this)

        battleByMarker = HashMap()
        circleByMarker = HashMap()
    }


    override fun onMapReady(map: GoogleMap) {
        info("map is ready")

        googleMap = map

        if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
            googleMap.isMyLocationEnabled = true
        } else {
            LocationService.askLocationPermission(this)
        }

        val settings = googleMap.uiSettings
        settings.isZoomControlsEnabled = true
        settings.isMapToolbarEnabled = false
        settings.isMyLocationButtonEnabled = true

        googleMap.moveCamera(CameraUpdateFactory.newLatLngZoom(viewPoint, viewZoom))

        googleMap.setOnInfoWindowClickListener { marker ->
            State.chosenBattle = battleByMarker[marker.id]
            BattleActivity.gotHereFrom = MapActivity::class.java
            startActivity<BattleActivity>()
        }

        googleMap.setOnMapClickListener { latLng ->
            if (visibleCircle != null) {
                visibleCircle!!.isVisible = false
                visibleCircle = null
            }
        }

        googleMap.setOnMarkerClickListener { marker ->
            if (visibleCircle != null) {
                visibleCircle!!.isVisible = false
            }
            visibleCircle = circleByMarker[marker.id]
            visibleCircle!!.isVisible = true
            false
        }

        placeMarkers()
    }

    private fun placeMarkers() {
        BattleManager.getAllAndDo(object : Callback<List<Battle>> {
            override fun onResponse(call: Call<List<Battle>>, response: Response<List<Battle>>) {
                val battles = response.body()
                for (nearBattle in battles!!) {
                    val newMarker = googleMap.addMarker(MarkerOptions()
                            .position(nearBattle.location)
                            .title(nearBattle.name)
                            .snippet(nearBattle.entriesCount.toString() + " photos"))
                    battleByMarker[newMarker.id] = nearBattle

                    val newCircle = googleMap.addCircle(CircleOptions()
                            .center(nearBattle.location)
                            .radius(nearBattle.radius.toDouble())
                            .strokeColor(Color.GREEN)
                            .strokeWidth(4f)
                            .visible(false))
                    circleByMarker[newMarker.id] = newCircle
                }
                info("placed markers for " + battles.size + " battles")
            }

            override fun onFailure(call: Call<List<Battle>>, t: Throwable) {
                Utils.showToast(this@MapActivity, "Failed to get battles, try to reload map.")
                error("cant get battles: $t")
            }
        })
    }

    public override fun onPause() {
        super.onPause()
        viewPoint = googleMap.cameraPosition.target
        viewZoom = googleMap.cameraPosition.zoom
    }

    fun goCreateBattle(v: View) {
        if (State.currentUser.battleCreationLimit == 0) {
            toast("Sorry, you've spent all of your battle creations for this week.")
            return
        }
        State.creatingBattle = true
        startActivity<CreateBattleActivity>()
    }

    override fun onBackPressed() {
        val intent = Intent(this, gotHereFrom)
        startActivity(intent)
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        if (requestCode == LocationService.REQUEST_LOCATION_PERMISSION) {
            if (grantResults.size != 1 || grantResults[0] != PackageManager.PERMISSION_GRANTED) {
                toast("If you decide to participate, grant this permission later in settings.")
                info("didn't get location permission")
            } else {
                info("got location permission")

                try {
                    googleMap.isMyLocationEnabled = true
                } catch (e: SecurityException) {
                    //try catch block is required by API 23 and higher, never get here
                }
            }
        } else {
            super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        }
    }

    companion object {
        const val DEFAULT_ZOOM = 14f
        val DEFAULT_VIEW_POINT = LatLng(59.930969, 30.352445)

        var gotHereFrom: Class<*>? = null

        var viewPoint = DEFAULT_VIEW_POINT
        var viewZoom = DEFAULT_ZOOM
    }
}
